### PR TITLE
MINOR: Fix gradle error writing test stdout

### PR DESCRIPTION
**Pull request by vvcephei**
Fix the following message we have been seeing in recent builds:

> ERROR: Failed to write output for test org.apache.kafka.streams.integration
> .QueryableStateIntegrationTest.shouldBeAbleQueryStandbyStateDuringRebalance
> java.lang.NullPointerException: Cannot invoke method write() on null object

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
Hey @ewencp @ijuma ,

We've been seeing the above error a lot in the build logs. While I'm not sure why, it seems like it must be that somehow, the `beforeTest` block doesn't execute before `onOutput`. There's already a comment in the `onOutput` block that suggests we might see output on TestDescriptors that hadn't been triggered in `beforeTest` already, so maybe a recent Gradle upgrade expanded the number of such cases?

I initially had a different idea, but it seems like we're comfortable discarding output that wasn't part of a "test", so I changed this fix just to directly check if `beforeTest` had run before `onOutput`.

WDYT?

**Comment by vvcephei** [comment]
For what it's worth, I ran the build locally, and observed one failing test, `SmokeTestDriverIntegrationTest. shouldWorkWithRebalance`.

I also observe exactly one test output file:
```
[john@arcturus kafka]$ find . -name *.test.stdout./streams/build/reports/testOutput
/org.apache.kafka.streams.integration.SmokeTestDriverIntegrationTest.shouldWorkWithRebalance.test.stdout
```

And it does contain the output from that test.

File: build.gradle
Comment (ewencp) [inline build.gradle:255]: The problem with just ignoring here is that we don't really have an explanation for how this can happen. I forget the exact details of what the output was (other than note that it might be "leftover" from buggy test runner, and my immediate thought that some could also be from things like static initialization). If anything, we actually might want to log *more info. Aside from potential log size (and noise) I don't remember why I didn't include the `toe.message` in the exception handler logging below.

Have we run with that modification to see if it's possible to tell what type of log output is getting lost? We could also change `ERROR` to `WARNING`. The main reason I wanted it at the scarier `ERROR` level was because it was unclear at the time just how frequent it would be, and so there was some indication in the log that some output from *some* test that was running before that one might have missing log data.

We also might not need the stack trace, which adds more noise. I think I included that so if we saw it consistently we could try to track down when this would commonly occur (e.g. if we could track it down to something like static initialization as the common cause).

Comment (vvcephei) [reply build.gradle:255]: Hi @ewencp , thanks for the review!

It wasn't clear to me that we think it shouldn't happen. With that knowledge, it does seem worthwhile to include some debugging logs to try and understand what is going on.

FWIW, I like the stacktrace. That's actually the thing that tipped me off that the problem was with the build itself, rather than a test, thereby saving me from hours of digging in the wrong direction.

I might be an idealist, but I'm not worried about the "level of noise", as in "number of log lines". The fact is that the build system isn't behaving the way we think it should, so we should try to figure out why. Once we understand why, we'll know what to do, and _then_ we won't have to listen to any more noise.

I'll update the PR.

Comment (ewencp) [reply build.gradle:255]: Yeah, unfortunately iirc it previously looked like Gradle's implementation of this feature was just buggy so might not be completely fixable without figuring out what to fix upstream. I prefer more info as well, but since people already often have trouble finding the real source of issues in long logs, I am a bit hesitant to make them even more voluminous. I think current modification is reasonable tradeoff.

**Review by ewencp — APPROVED** [review]

**Comment by vvcephei** [comment]
Ok, that last run was pretty noisy, even for me :)

Here's a random excerpt from the logs:
```
01:58:16.865 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,668] INFO [Controller id=0] Resigned (kafka.controller.KafkaController:66)
01:58:16.865 ].
01:58:16.865 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,668] INFO [ZooKeeperClient Kafka server] Closing. (kafka.zookeeper.ZooKeeperClient:66)
01:58:16.865 ].
01:58:16.964 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,769] INFO EventThread shut down for session: 0x1007f9c06630000 (org.apache.zookeeper.ClientCnxn:524)
01:58:16.964 ].
01:58:16.964 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,769] INFO Session: 0x1007f9c06630000 closed (org.apache.zookeeper.ZooKeeper:1422)
01:58:16.964 ].
01:58:16.964 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,770] INFO [ZooKeeperClient Kafka server] Closed. (kafka.zookeeper.ZooKeeperClient:66)
01:58:16.964 ].
01:58:16.964 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,770] INFO [ThrottledChannelReaper-Fetch]: Shutting down (kafka.server.ClientQuotaManager$ThrottledChannelReaper:66)
01:58:16.964 ].
01:58:16.964 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,795] WARN [Consumer clientId=instance-0-StreamThread-1-consumer, groupId=lag-fetch-shouldFetchLagsDuringRestoration] Connection to node 0 (localhost/127.0.0.1:38973) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient:762)
01:58:16.964 ].
01:58:17.064 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,890] WARN [AdminClient clientId=instance-0-admin] Connection to node 0 (localhost/127.0.0.1:38973) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient:762)
01:58:17.064 ].
01:58:17.165 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:56,953] WARN [AdminClient clientId=adminclient-91] Connection to node 0 (localhost/127.0.0.1:33189) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient:762)
01:58:17.165 ].
01:58:17.165 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:57,003] INFO [ThrottledChannelReaper-Fetch]: Stopped (kafka.server.ClientQuotaManager$ThrottledChannelReaper:66)
01:58:17.165 ].
01:58:17.165 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:57,003] INFO [ThrottledChannelReaper-Fetch]: Shutdown completed (kafka.server.ClientQuotaManager$ThrottledChannelReaper:66)
01:58:17.165 ].
01:58:17.165 WARNING: unexpectedly got output for a test org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 22:10:57,004] INFO [ThrottledChannelReaper-Produce]: Shutting down (kafka.server.ClientQuotaManager$ThrottledChannelReaper:66)
```

I think I'll put your original condition back to be silently ignored, and keep the warning for my new condition. 

I did occur to me while looking at the printed warnings that the messages maybe aren't being printed before `beforeTest` is called, but instead after `afterTest` is called. I.e., some threads spawned during the test have outlived the test itself and continued to print to the output stream. It might have already been obvious to you, but I was only thinking that these rogue outputs had been made without `beforeTest` being executed first, but this case is indistinguishable from `beforeTest ... test ... afterTest ... then more output`.

**Comment by vvcephei** [comment]
Ok, after making the original condition just silently drop output, I see one mighty interesting warning:

```
WARNING: unexpectedly got output for a test null.Gradle Test Executor 8 that we didn't previously see in the beforeTest hook. Message for debugging: [[2020-02-25 23:39:57,683] WARN [AdminClient clientId=adminclient-14] Connection to node -1 (localhost/127.0.0.1:8080) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient:762)
].
```

I.e., the descriptor's className was `null` and its name was `Gradle Test Executor 8`... Uh, what?

**Comment by vvcephei** [comment]
Ok @ewencp , I've added a condition to silently ignore the case where the className is `null`, since it looks like that last output was just leaked totally outside of a test.

It actually has me mildly concerned that we have threads hanging around indefinitely during the build, and wondering if it's worth trying to make sure that none of the threads spawned during a test outlive the test...

Then again, I think I'm about done bikeshedding this at the moment. Do you feel good enough about this as an improvement on the status quo for now?

**Review by ewencp — APPROVED** [review]
Yeah, agreed I think this is fine for now. It feels a bit wrong since we clearly don't get entirely correct results, but since the goal here is to just aid in debugging, unless we see it causing problems in addressing that need I think we can make this tradeoff.

The exact nature of the issue was difficult to pin down previously, so not too surprised it's still unclear. Whether it is before the test or after is unclear, made worse by this case where the class name isn't even filled in. I suspect there might be some code that isn't properly synchronized or something. I think the reason I didn't think it as during afterTest was because the output didn't seem related to the test at all (even if it had been reported late). Again, been awhile, but I vaguely remember seeing output that appeared to possibly come from a test that had been logged just before, but was attributed to the next test. But since gradle is giving us questionable data, I think it's fine to just do the best we can with it.

re: leaking threads, I could have sworn we had some sort of leak checking, but since I can't seem to find it, I guess maybe we just did some of those checks ad hoc, but not permanently, during some PRs or something. It would be a nice thing to validate -- at least logging at the end of tests seems like it'd be a good starting point. Unfortunately, verifying pre & post test won't work with test parallelism so I'm not sure how helpful it would be in actually tracking down broken tests unless the follow up had to be a serialized run of tests.

In any case, let's merge this, and if you feel motivated to add some checks on threads, would be happy to review those too :)

Oh, and I'll leave you to merge -- I'm not sure which set of branches makes sense to backport to. It sounds like maybe this is only recent, but it also probably doesn't hurt to just backport to whichever branch this code was initially added to.

**Comment by ijuma** [comment]
We have thread leak detection in one of the kafka integration test harness classes. I think it's the ZK one.

**Comment by vvcephei** [comment]
Thanks, @ewencp !

I'll keep an eye on the log messages and mull over what else we can do here. I guess one hope I'm holding onto is that, from the perspective of debugging failed tests, maybe the "evidence" that we need is in the middle of the output for that test. So even if there's some messages going awry at the edges, hopefully, they don't matter for debugability.

Anyway, merging and cherry-picking now...

**Comment by vvcephei** [comment]
Thanks, @ijuma ! I'll look for it, and see if I can apply the approach more generally.

**Comment by vvcephei** [comment]
Cherry-picked to 2.3 (the first version that had the onOutput hook).

Checking with the release managers before cherry-pick to 2.4 and 2.5...

**Comment by vvcephei** [comment]
Cherry-picked to 2.4, waiting on the 2.5.0 release before cherry-picking to 2.5

**Pull request merged**
